### ISSUE TITLE
RFE: set builsubdir to the *first* extracted archive not the last one

### ISSUE THREAD

**Issue by nim-nim** [rpm-software-management/rpm#551]
rpm convention is to put the main archive in SOURCE0 and ancilliary sources in SOURCEX
(this way SOURCE0 is sure to exist and be stable, and can be aliased to SOURCE, while the number of addons and their index can vary over a package time)

However rpm does not respect its own convention, sets builsubdir to the *last* processed archive, and cds to this last archive at the start of build

This has the following deleterious side effects:

1. the packager ends up working in an addon subdirectory, which is utterly confusing, because this is not the subdirectory he cares about most, and will change over time depending on the number of addon archives he adds to his spec

2. you can not do any cleanup of the extracted archives in %prep, that involves things that should end up in the packager main work directory, because this directory does not exist before all archives have been extracted. Instead of the natural:
```specfile
%prep
for each SOURCEX
  %setup
  cd to extracted archive
  archive prep work
done
```
you need to code *two loops*
```specfile
%prep
for each SOURCEX
  %setup
  remember the extracted dir
done ← at this point only builsubdir is set to its definite value
for each extracted dir
  cd extractdir
  do prep work that involves the final workdir
done
```
This is needlessly byzantine, just set `builsubdir` to the first archive subdirectory, so it is garanteed to exist after the first %setup invocation, and packagers get warped to their main project files at the start of %build

**Comment by n3npq** [comment]
The %setup macro-in-c was never intended to be used multiple times: see the shell code generated, and look at the needless unnecessary commands generated if/when %setup is used multiple times.

You are likely better off (for whatever you are attempting) using %setup once to set %_buildsubdir, or even not using %setup at all, and writing the simple shell commands to unpack archives however you wish, rather than complaining about the (known since last century) Byzantine behavior of %setup

**Comment by ascherer** [comment]
I believe that's what the `-n` option of `%setup` is for. You can (or have to) set/repeat the desired `%buildsubdir` on every invocation of `%setup`. (IIRC, multiple `%setup`s are deemed esoteric/byzantine anyway.)

**Comment by nim-nim** [comment]
The problem is not setting it. The problem is that rpm uses the last extracted directory as default workdir, preventing any preparation that involves this directory before every archive has been extracted

**Comment by Conan-Kudo** [comment]
You're supposed to use `%setup` once, and tell it how many archives it needs to extract before it it should change into the workdir.

For example: `%setup -q -b 3` tells it to unpack `Source0`, `Source1`, `Source2`, and `Source3` before changing into the `%name-%version` workdir. If you want another directory, you need to override with `-n`.

**Comment by nim-nim** [comment]
I don't doubt that's what the setup authors intended, but really, it's overly clever for human beings, that like their commands simple and easy to remember, it's overly clever for scripts, that like to loop over the same command not compute complex lists of arguments, it's overly clever for whoever is tasked to explain it so anyone else, and besides, it does not handle patching of intermediate sources well.

Which is why from time immemorial, starting with the heroic period when maximum rpm was written, the documented and actual `%setup` use was not what the setup authors intended, but one setup call per archive.

So could we make the way people actually use `%setup` work better please?

**Comment by n3npq** [comment]
You are absolutely correct that %setup from time immemorial has been overly complicated, non-intuitive, and *documented*.

Expecting users to change their behavior 20y later is a political task, better left to distro packaging committees and package reviews: %setup does what %setup does.

Meanwhile you critiquing an implementation 20y later in multiple RFE's because it doesn't fit your obscure usage case isn't helpful either.

If %setup is not appropriate for whatever you are attempting, then

DO NOT USE %setup

It's really that simple.

**Comment by n3npq** [comment]
FWIW, here is an implementation that would permit first rather than last definition of %_buildsubdir to be used under user control.

RPM5 permits read only macros which -- when set -- prevent a macro value from being changed.

The syntax prefixes a macro name with a '.' in order to set (actually toggle, but let's not go there) a read only flag.

The '.' is not part of the name, is stripped off to set a read only flag. Any attempt to redefine the macro value is ignored (and warned iirc).

The entire implementation is a couple dozen lines of code.

Adding another configurable macro to define whether _buildsubdir or its read only variant with a '.' prefix is to be used can be configured in rpm in multiple ways.

The read only variant will of course be the first, rather than the last, value defined by %setup, entirely satisfying the RFE stated herein.

**Comment by n3npq** [comment]
Even simpler -- only one line of code needs to change -- would be to define _buildsubdir only if it is not already defined.

That might even permit overriding _buildsubdir to be set from the spec file, or -- with Yet Another POPT alias -- from the rpmbuild command line.

And of course %setup becomes even more complex and counter intuitive because the behavior then starts to be convolved with how a spec file is written, or how rpmbuild is invoked.

And %setup then starts to diverge from documented 20yo behavior.

(aside)
I recommend these solutions merely to shut you up and watch this RFE be closed: I'm tired of your pontifications about rpm's unfriendly user behavior that do not include PR's.

**Comment by pmatilai** [comment]
Part of the problem is the lengths people go to in order to use %setup for unpacking sources even when it ends up being counter-beneficial. Just call it once to setup the main directory and do the rest with tar/unzip/whatever. It's not like the magic %setup incantation required to unpack dozen sources in just the right combo is going to be somehow easier to understand than doing the same manually as a script.

Another point is that indeed, if it's not supposed to be used more than once then rpm should prevent that, or at least warn about it. Telling people "it's not supposed to be used that way" doesn't scale.

**Comment by nim-nim** [comment]
Panu, one huge reason `%setup` is still a mess after all this time it that it is linked to the `SourceX` black magic. And yes rpmbuild will complain loudly if you deviate from the way it thinks things should be done. And yes it will refuse to process the resulting spec. And the documentation is piss-poor. 

There are layers over layers of workarounds in `SourceX` and `%setup` just to avoid admitting parts of the original rpm design didn't work out in practice. So you have to shoehorn archive names behind an artificial # in `SourceX` (because admitting the file name is not always part of the URL and a separate tag is needed is out of the question), you have the way distros have been forced to invent %dist (making Release, in effect, two-tags-in-one), you have the way `%setup` still tries to chainbuild, when maximum rpm, written in *2000*, already document the way people actually use it and expect it to work is multiple `%setup` calls and so on I probably forget half of it.

So it's not surprising people just use `%setup` by rote, just to get past this ordeal as fast as possible, and get to the parts of the spec where they can do things without rpm complaining and trying to impose unintuitive overcomplicated misdesigns on them.

Just provide in rpm a `%setup` alternative that just does what you propose (unpack one archive, without esotheric chaining rules to avoid writing one more extraction command, and with simple ways to query or set buildsubdir that rpm won't stomp on later), and you'll see `%setup` use and misuse drop quickly.

Or, alternatively, fix `%setup` to work how people expect it to work, without the magic they don't expect and don't need.

But having sour grapes about people using `%setup` the way maximum rpm documented 18 *years* ago? And trying to teach them a lesson in the correct way to do things via warnings? That will end badly for everyone involved.

**Comment by n3npq** [comment]
First of all, MaximumRPM was written in 1997, not 2000.

Second, the vast majority of packages have no need for a fancier %setup, nor enhancements like %autosetup. Most software has been distributed on the Internet as a tarball that can be downloaded, and most packages do not need anything more than the few lines of shell script that %setup generates to simplify packaging. Sure times have changed since last century: but surely you are not going to question that the dominant paradigm everywhere is a single tar ball, downloaded, and packaged, where %setup is quite adequate to do the job.

Third, bringing up issues like how URL's are represented in a SourceN: directive, and the parsing quirks in rpmbuild, has nothing whatsoever to do with %setup. RPM is not used as a downloader in production at any time this century. Nor is there any code in rpm-4.14 that can parse a URL: a string is passed to curl (or wget) which does the job, nothing more.

Fourth, positing that %setup should work as "people expect it to work" is pure sophistry. Packagers expect %setup to do what it has always done, across 25 years of spec files and packaging. Every change to RPM has a huge ripple effect, the rate of adoption of any change is measured in years or decades, because of promises of long term support in enterprise distro's like RHEL and CentOS.

If you claim otherwise, or think it's time for a change, then by all means get off your bully pulpit and

Show me (and all of us) your code.

**Comment by ffesti** [comment]
@n3npq It looks like you are deleting most of your comments after posting them here. This disrupts the discussion and renders the tickets unusable. Also note that they are still available in the mailing-list archive and everyone's mail folder - if you don't want your words to be archived and available for future review please post them else where.

**Comment by pmatilai** [comment]
>  Just provide in rpm a %setup alternative that just does what you propose (unpack one archive, without esotheric chaining rules to avoid writing one more extraction command, and with simple ways to query or set buildsubdir that rpm won't stomp on later), and you'll see %setup use and misuse drop quickly.

@nim-nim, but isn't that just about exactly what %setup by default does? Calling it once with no fancy arguments will unpack the first source and define %_buildsubdir, and from there on you can do whatever you want with the rest. Or call it just to create a directory of your choosing and do the rest manually. 

Like said rpm could do better there, and having different semantics for %_buildsubdir is trivial, but if calling %setup multiple times hurts and makes things unnecessarily complicated, then don't!

**Comment by n3npq** [comment]
@festi: no.

**Comment by nim-nim** [comment]
> @nim-nim, but isn't that just about exactly what %setup by default does? 

Not really, since the chainbuilding interferes `%setup` reuse with the next archive

People just want to align
```specfile
%somesetup -z 1
%somesetup -z 2
%somesetup -z 3
```
Like they align path or source declarations

And they want to have some simple way to set the primary root where `%build` will chdir and various other macros like autodeps will look for generated files.

They do *not* want to use one macro for `Source0` and switch to something else for other sources.
They do *not* want this simple stupid natural pattern to have weird side effects because the expansion macro was really designed for a chainbuilding that almost not one actually uses or need.  

> Calling it once with no fancy arguments will unpack the first source and define %_buildsubdir,
No it won't, the variable is not really set before %build and rpm will overwrite any attempt to set it manually.

**Comment by n3npq** [comment]
RPM has no concept of "chain building": there is only one build in the spec file, and there is only one SRPM.

Please define whatever terms you choose to make up before using so that others might understand your complaints.

**Comment by pmatilai** [comment]
Oh and BTW, if the reason for wanting to use %setup "at any cost" is that it knows how to extract tar, zip, ruby gems and whatnot automatically, then *that* might be another concrete something we could look into, ie add a simpler lower level %extract or such helper macro/tool that can do what %setup can, but only for one source, and without the side-effects.

**Comment by nim-nim** [comment]
> > No it won't, the variable is not really set before %build
> 
> NOW we're getting somewhere. %setup parsing certainly defines it, and this happens before any shell script is executed, but I see it gets lost somewhere in the cracks of %prep parsing. I can take a look at it, all I remember off-hand is that %prep parsing is entirely different from the rest of the build-sections.
> 
> However I still don't see the big problem here. %buildsubdir is not some magic random value calculated by rpm, it's always %{name}-%{version}, or literally whatever you pass to %setup with -n argument. It's also the directory you'll find yourself in the shell, so derivable from "basename ${PWD}" inside %prep.

The packaging pattern is

```specfile
Source0: main-archive
Source1: addon1
Source2: addon2
Source3: addon3

[...]
%prep
%setupsomething -z 0 (main archive)
%setupsomething -z 1 (addon 1)
%setupsomething -z 2 (addon 2)
%setupsomething -z 3 (addon 3
```
It has the nice properties that:
1. it is simple and easy to understand
2. it is simple and easy to explain
3. it is patch and preparation friendly (just put the `%patch` or any other archive fixup directive after the corresponding `%setupsomething` line)
4. it is spec-maintenance friendly (want to remove a particular addon ? just remove the corresponding lines, you end with a simple diff with added and removed lines, without side-effects on other lines – or changed lines if you prefer to call it that way)

And I used  `-z` instead of the horrible `-a -b -T -n` mix that no one understands at first and is only there to permit chaining and to avoid standardizing a way to specify the topdir upstream uses with its archives, and the topdir the packager wishes to generate (if uptream is topdir-less).

That's the packaging pattern Maximum rpm already documented in the last century, and having spent some months writing macros which are intended to be used in `%prep`, and tried to document them to others, I understand 100% why the Maximum rpm authors chose to document it that way (and I agree with them).

Maximum rpm  is the main vulgarisation document that attemps to explain how to use `%setup` with multiple archives. I'd say it succeeds poorly, in large part due to `%setup` quirks, but at least it tries.

Now, as Neal explains, those quirks are largely due to the historical `%setup` chaining design. But almost nobody uses `%setup` that way. First, it is not documented in any specfile writing vulgarization document. And second, even it it *was* documented, it does not have the nice packager-friendly properties of the no-chaining approach which *is* documented in Maximum rpm.

However, even even if you use `%setup` without chaining, you still get bitten by the chaining design.

Because `%setup` assumes you will use it in one call, it makes no provision to set `%{buildsubdir}` cleanly. `%{buildsubdir}` is whatever you passed `%setup` with -n, why you use several `%setup` calls, how weird, why are you doing that? (I've already explained why).

So packagers mostly accept this is dark rpm vodoo and that rpm will chdir to the last addon directory in `%build` (which is definitively *not* their main work directory).

This RFE is about setting `%{buildsubdir}` by default to something sensible: the extraction directory of the `Source0` archive, which will almost always be the main archive processed by the spec, due to historical conventions.

Alternatively, just making it possible to override in `%prep` the mess `%setup` makes of this all, works for me too.

I'm *definitely* not interested in anything that would be equivalent to

```specfile
%prep
%myprepmacro

%build
%fixmyprep ← workaround %{buildsubdir} selection in %prep
```

**Comment by n3npq** [comment]
@nim-nim:

Your "nice properties" arguments are not persuasive.

Point by point:

1. Simple and easy to understand

For example, the common case of a single well formed compresset tarball with a patch  in a spec file is even simpler than your design:
 
%prep
%setup
%patch

2. It is simple and easy to explain

The example I just gave is trivial to explain

3. It is patch and preparation friendly

Depends entirely on your POV. Rpm was intended for simple packages that can be independently upgraded. Rather than trying to do a complex "chained" (actually serial unpacking, not chained builds) build, history has shown that decomposing into smaller pieces that are simpler to upgrade is quite efficient, more efficient than "make world" type builds.

4. it is spec maintenance friendly

The most often reported problem with spec maintenance these days is the cost of editing thousands of spec files, not how easy it is to "maintain" spec files. While it is certainly easy to delete lines Han it is to, say, change a macro value by editing, the drudgery of changing hundreds or thousands of files in order to conform to some Newer! Better! Bestest! Syntax or policy.

**Comment by nim-nim** [comment]
> For example, the common case of a single well formed compresset tarball with a patch in a spec file is even simpler than your design:
> 
> %prep
> %setup
> %patch

And that can be trivialy declined in

```specfile
%prep
%setup -z 0
%patch0
%setup -z 1
%patch1
%setup -z 2
%patch2
```

(of course one could genereate `%setupX` names too, that's another can of worms created by some of historical rpm choices).

But anyway I'm not here to discuss why single archive is simpler than multiple archives. That's self-evident. However some people **do** want to process multiple archives. Being able to handle multiple archive **was** the first request of Fedora's rpm maintainers when I posted my first macro iterations months ago. So bzzzt people wanted me to look at it, so I looked at it. And it was a major PITA due to many rpm quirks that ended up as RFEs here.

My personal use cases do not necessarily match the use cases of everyone in the distribution or its derivatives.

> of spec files, not how easy it is to "maintain" spec files. While it is certainly easy to delete lines Han it is to, say, change a macro value by editing, the drudgery of changing hundreds or thousands of files in order to conform to some Newer! Better! Bestest! Syntax or policy.

I'm not in the business of setting policies, this is free software, people will use the result if they find it useful and convenient, which is why I make damn sure it is useful and convenient, because I have no interest in wasting my efforts on things almost no one uses in reality, or in explaining to the users of my macros they are idiots that should just do unnatural and unintuitive xxx due to some rpm historical implementation quirk they have no interest in (and just want to be fixed).

**Comment by n3npq** [comment]
You have chosen rpm spec files and macros for your implementation.

Neither was designed for multiple builds in a single package, nor was rpm for that matter.

Instead of complaining about rpm deficiencies and quirks, you might wish to rethink your approach, look at the inputs of archive/patch names, choose whatever directory hierarchy you wish, and script up your "chained builds" in whatever programming language you choose.

Then run your "chained builds" within an rpm spec file.

**Comment by pmatilai** [comment]
This could potentially be solved by #860 (ie just allow macros to do what they want with it)

**Comment by nim-nim** [comment]
Thanks, I need to investigate if #860 solves the problem for me

**Comment by pmatilai** [comment]
FWIW, #1976 moves all the `%setup` extraction logic into an external `rpmuncompress` command, so you can just call that on any archive format rpm knows about and never mind the bollocks with `%setup`. And since `%buildsubdir` is now (since 6caca84c904423fbe75674c1589f3457c4927281) a regular macro, you really don't *need* to muck with %setup at all.

**Comment by ffesti** [comment]
With the latest changes to the `%setup` macro this may now be easier and should be looked at again.
